### ISSUE TITLE
Attachments not working properly in latest version

### ISSUE THREAD

**Issue by Ben-MyBB** [mybb/mybb#3618]
Hi everyone, I am pasting a portion of this here:

https://community.mybb.com/thread-219817.html

First in this file:

https://github.com/mybb/mybb/blob/featur...upload.php

The function:

function add_attachments(&#36;pid, &#36;forumpermissions, &#36;attachwhere, &#36;action=false)

looks quite deficient and buggy. It's error reporting is quite bad. Should complaint when some of the things fail but it silently ignores the errors. Very badly programmed (no alternative else reporting error).

I made some amendments (in my copy) checking for &#36;total == 0 that is what was happening and realized the error was reported but not displayed.

I then realized the form was sending "attachment" instead of "attachments" (notice the ending S) so I wen ahead and fixed the template to change the name to "attachments".

But then some other error is happening after the "fix". Of course no error is displayed for whatever reason.

Any Idea on why this error is not displayed only showing an empty alert when this variable is populated:

&#36;ret['errors'][] = ...

Maybe I could track down this further.

Ok I figured out everything and I found some bugs in MyBB 1.8.19.

First seems the upgrade path using changed files and update script misses to add a new template piece called "error_inline_item"

in inc/functions.php (https://github.com/mybb/mybb/blob/87e187...ctions.php) the function 

function inline_error(&#36;errors, &#36;title="", &#36;json_data=array())

I reverted the line:

eval("\&#36;errorlist .= \"".&#36;templates-&gt;get("error_inline_item")."\";");

to:

&#36;errorlist .= "&lt;li&gt;".&#36;error."&lt;/li&gt;\n";

as in here https://github.com/mybb/mybb/blob/b7ffd9...ctions.php

That way I could print the errors. So then I got this cryptic error:

The "C" file is empty

Where C was the first letter of the file I attempted to upload.

So this made me take a look at parts of the code to figure out from where that could come.

Long story short there is this bug in the new uploads code with that index

in file inc/functions_upload.php

/**
 * Process adding attachment(s) when the &quot;Add Attachment&quot; button is pressed.
 *
 * @param int $pid The ID of the post.
 * @param array $forumpermission The permissions for the forum.
 * @param string $attachwhere Search string &quot;pid='$pid'&quot; or &quot;posthash='&quot;.$db-&gt;escape_string($mybb-&gt;get_input('posthash')).&quot;'&quot;
 * @param string $action Where called from: &quot;newthread&quot;, &quot;newreply&quot;, or &quot;editpost&quot;
 *
 * @return array Array of errors if any, empty array otherwise
 */
function add_attachments($pid, $forumpermissions, $attachwhere, $action=false)
{
	global $db, $mybb, $editdraftpid, $lang;

	$ret = array();

	if($forumpermissions['canpostattachments'])
	{
		$attachments = array();
		$fields = array ('name', 'type', 'tmp_name', 'error', 'size');
		$aid = array();

		$total = isset($_FILES['attachments']['name']) ? count($_FILES['attachments']['name']) : 0;
		$filenames = &quot;&quot;;
		$delim = &quot;&quot;;
		
		//if(!isset($_FILES['attachments']['name']))
		//{
		//	$ret['errors'][] = &quot; no files &quot;;
		//}
		
		if(0 == $total)
		{
			$ret['errors'][] = &quot; no files &quot;;
		}
		
		for($i=0; $i&lt;$total; ++$i)
		{
			foreach($fields as $field)
			{
				$attach1[$field] = $_FILES['attachments'][$field][$key];
				$attachments[$i][$field] = $_FILES['attachments'][$field]; //[$i]; &lt;----- nasty bug
			}

			$FILE = $attachments[$i];
			if(!empty($FILE['name']) &amp;&amp; !empty($FILE['type']) &amp;&amp; $FILE['size'] &gt; 0)
			{
				$filenames .= $delim . &quot;'&quot; . $db-&gt;escape_string($FILE['name']) . &quot;'&quot;;
				$delim = &quot;,&quot;;
			}
		}
		
		

		if ($filenames != '')
		{
			$query = $db-&gt;simple_select(&quot;attachments&quot;, &quot;filename&quot;, &quot;{$attachwhere} AND filename IN (&quot;.$filenames.&quot;)&quot;);

			while ($row = $db-&gt;fetch_array($query))
			{
				$aid[$row['filename']] = true;
			}
		}

		foreach($attachments as $FILE)
		{
			if(!empty($FILE['name']) &amp;&amp; !empty($FILE['type']))
			{
				if($FILE['size'] &gt; 0)
				{
					$filename = $db-&gt;escape_string($FILE['name']);
					$exists = $aid[$filename];

					$update_attachment = false;
					if($action == &quot;editpost&quot;)
					{
						if($exists &amp;&amp; $mybb-&gt;get_input('updateattachment') &amp;&amp; ($mybb-&gt;usergroup['caneditattachments'] || $forumpermissions['caneditattachments']))
						{
							$update_attachment = true;
						}
					}
					else
					{
						if($exists &amp;&amp; $mybb-&gt;get_input('updateattachment'))
						{
							$update_attachment = true;
						}
					}

					$attachedfile = upload_attachment($FILE, $update_attachment);

					if(!empty($attachedfile['error']))
					{
						$ret['errors'][] = $attachedfile['error'];
						$mybb-&gt;input['action'] = $action;
					}
				}
				else
				{
					$ret['errors'][] = $lang-&gt;sprintf($lang-&gt;error_uploadempty, htmlspecialchars_uni($FILE['name']));
					$mybb-&gt;input['action'] = $action;
				}
			}
		}
	}
	else
	{
		$ret['errors'][] = &quot;no permission&quot;;
	}
	

	return $ret;
}
This fixes it for me at least for ONE file upload (and should not be correct for multiple uploads)

I think maybe a better implementation for multiupload would be this (is not working for me in php 7 the hardcoded index)

&lt;form action=&quot;&quot; method=&quot;post&quot; enctype=&quot;multipart/form-data&quot;&gt;
&lt;p&gt;Pictures:
&lt;input type=&quot;file&quot; name=&quot;pictures[]&quot; /&gt;
&lt;input type=&quot;file&quot; name=&quot;pictures[]&quot; /&gt;
&lt;input type=&quot;file&quot; name=&quot;pictures[]&quot; /&gt;
&lt;input type=&quot;submit&quot; value=&quot;Send&quot; /&gt;
&lt;/p&gt;
&lt;/form&gt;

&lt;?php
foreach ($_FILES[&quot;pictures&quot;][&quot;error&quot;] as $key =&gt; $error) {
    if ($error == UPLOAD_ERR_OK) {
        $tmp_name = $_FILES[&quot;pictures&quot;][&quot;tmp_name&quot;][$key];
        // basename() may prevent filesystem traversal attacks;
        // further validation/sanitation of the filename may be appropriate
        $name = basename($_FILES[&quot;pictures&quot;][&quot;name&quot;][$key]);
        move_uploaded_file($tmp_name, &quot;data/$name&quot;);
    }
}
?&gt;

as in here:

http://php.net/manual/en/features.file-u...method.php

Original thread: [Attachments not working properly in latest version](https://community.mybb.com/thread-219865.html)

**Comment by waldoalvarez00** [comment]
Thanks for adding the issue. I was going to open it. I am adding my exact php version:

# php --version
PHP 7.0.32 (cli) (built: Sep 13 2018 16:50:42) ( NTS )
Copyright (c) 1997-2017 The PHP Group
Zend Engine v3.0.0, Copyright (c) 1998-2017 Zend Technologies

**Comment by waldoalvarez00** [comment]
I was expecting this nasty bug to be fixed in next release since I have two choices here. Skip security updates or download and reprogram the forum core code with each update in order to have the so much needed functionality. I cannot downgrade the php version just because some other software running in the server requires it.

**Comment by AniLeo** [comment]
I'm still experiencing some seemingly random occasions where users aren't able to upload attachments. I haven't figured any trigger to the issue, so I can't provide steps to reproduce it. Are there any news regarding this issue?

**Comment by tinuva** [comment]
Any update on this?

**Comment by JordanMussi** [comment]
I think this is possibly duplicated by #3886 and #3926 has been proposed to fix it.

**Comment by lairdshaw** [comment]
I think it's actually a unique issue, @JordanMussi. I came across it (at least, I think it's the same issue) independently not long ago. The issue as I understand it arises in the conditional on [line 748](https://github.com/mybb/mybb/blob/842d558cfe3e963ffafc2bc9e6b0c27d5bf62ed1/inc/functions_upload.php#L748) of inc/functions_upload.php. I have seen fit to replace that line with this code:

		// If there was PHP file upload error other than an empty file, then
		// avoid the checks below that might result in
		// upload_attachment() not being called, because that function
		// handles such errors.
		if ($FILE['error'] && $FILE['error'] != UPLOAD_ERR_NO_FILE) {
			$attachedfile = upload_attachment($FILE, $update_attachment);
			if (!empty($attachedfile['error'])) {
				$ret['errors'][] = $attachedfile['error'];
			}
		} else if (!empty($FILE['name']) && !empty($FILE['type'])) {

More specifically, the code as it is fails to respond to such PHP-generated errors as the file being larger than the `upload_max_filesize` php.ini setting. As it is (from memory - I would have to check to confirm), when a file is uploaded which is larger than the `upload_max_filesize` php.ini setting but less than the per-usergroup maximum upload filesize setting, no error is reported, but nor does the file upload succeed.

I'm happy to provide more details and/or to investigate deeper - am only reporting on my findings from several weeks ago. And the code that I've provided above could probably be refined. This function and its related functions could probably do with a bit of refactoring.
